pkg/services/sqlstore/stats.go

```diff
@@ -192,23 +184,118 @@ func GetSystemUserCountStats(ctx context.Context, query *models.GetSystemUserCou
-	activeUserDeadlineDate := time.Now().Add(-activeUserTimeLimit)
-	sb := &SqlBuilder{}
+func GetUserStats(query *models.GetUserStatsQuery) error {
+	err := updateUserRoleCountsIfNecessary(query.MustUpdate)
+	if err != nil {
+		return err
+	}
 
-	sb.Write(`SELECT `)
-	sb.Write(`(SELECT COUNT(*) FROM `+dialect.Quote("user")+` WHERE last_seen_at > ?) AS active_users,`, activeUserDeadlineDate)
-	sb.Write(roleCounterSQL("Viewer", "viewers", true)+`,`, activeUserDeadlineDate)
-	sb.Write(roleCounterSQL("Editor", "editors", true)+`,`, activeUserDeadlineDate)
-	sb.Write(roleCounterSQL("Admin", "admins", true)+``, activeUserDeadlineDate)
+	if query.Active {
+		query.Result = userStatsCache.active
+	} else {
+		query.Result = userStatsCache.total
+	}
 
-	var stats models.ActiveUserStats
-	_, err := x.SQL(sb.GetSqlString(), sb.params...).Get(&stats)
+	return nil
+}
+
+func updateUserRoleCountsIfNecessary(forced bool) error {
+	memoizationPeriod := time.Now().Add(-userStatsCacheLimetime)
+	if forced || userStatsCache.memoized.Before(memoizationPeriod) {
+		err := updateUserRoleCounts()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type memoUserStats struct {
+	active models.UserStats
+	total  models.UserStats
+
+	memoized time.Time
+}
+
+var (
+	userStatsCache         = memoUserStats{}
+	userStatsCacheLimetime = 5 * time.Minute
+
+	// 5 million users max cap is arbitrarily chosen to prevent this
+	// from taking far too much time.
+	userStatsPageLimit = 100
+	userStatsPageSize  = 50000
+)
+
+func updateUserRoleCounts() error {
+	type userRoles struct {
+		ID         int64 `xorm:"id"`
+		LastSeenAt time.Time
+		Role       models.RoleType
+	}
+
+	activeUserDeadline := time.Now().Add(-activeUserTimeLimit)
+
+	memo := memoUserStats{memoized: time.Now()}
+
+	_, err := x.Transaction(func(session *xorm.Session) (interface{}, error) {
+		for offset := 0; offset < userStatsPageLimit; offset += 1 {
+			usersSlice := []userRoles{}
+			err := session.SQL(`SELECT u.id, u.last_seen_at, org_user.role
+    FROM (SELECT id, last_seen_at FROM user LIMIT ? OFFSET ?) AS u LEFT JOIN org_user ON org_user.user_id = u.id
+    GROUP BY u.id, org_user.role;`, userStatsPageSize, offset*userStatsPageSize).Find(&usersSlice)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(usersSlice) == 0 {
+				break
+			}
+
+			activeMap := make(map[int64]models.RoleType, userStatsPageSize)
+			totalMap := make(map[int64]models.RoleType, userStatsPageSize)
+
+			for _, user := range usersSlice {
+				if current, exists := totalMap[user.ID]; exists {
+					if current == models.ROLE_ADMIN {
+						continue
+					} else if current == models.ROLE_EDITOR && user.Role == models.ROLE_VIEWER {
+						continue
+					}
+				}
+
+				if user.LastSeenAt.After(activeUserDeadline) {
+					activeMap[user.ID] = user.Role
+				}
+				totalMap[user.ID] = user.Role
+			}
+
+			memo.active = mapToStats(memo.active, activeMap)
+			memo.total = mapToStats(memo.total, totalMap)
+		}
```

Comment:
We could have a debug message if last `len(usersSlice) == userStatsPageSize` that most probably the `userStatsPageLimit` has been reached?

Reply:
I'll add that!